### PR TITLE
fix(bpdm-orchestrator): introduced pagination for handling pending and retention timeouts

### PR DESCRIPTION
**Pull request by SujitMBRDI**
## Description

This pull request introduces pagination for handling pending and retention timeouts in the orchestrator. By processing tasks in manageable batches, the changes prevent out-of-memory errors during the execution of large volumes of tasks, improving the system's stability and efficiency.

## Pre-review checks

Please ensure to do as many of the following checks as possible, before asking for committer review:

- [X] DEPENDENCIES are up-to-date. [Dash license tool](https://github.com/eclipse/dash-licenses). Committers can open IP issues for restricted libs.
- [X] [Copyright and license header](https://eclipse-tractusx.github.io/docs/release/trg-7/trg-7-02) are present on all affected files

### CODE REVIEW

**Review by nicoprow — COMMENTED** [review]
The logic looks sound to me but I think using such an unspecifc public function in the golden record task service in which you can basically place in any fetch and process function to be executed seems to be quite over the top. Especially when we consider that the TimeoutProcessBatchService does have two very similar methods that just differentiate in between which they give delegates as parameters to that function in the golden record task service.

Why not just create two separate functions in the golden record task service, one for fetching and processing pending timeouts, one for fetching and processing retention timeouts. If you then want to reduce duplicate code you could create a private function for batch processing these functions.

File: bpdm-orchestrator/src/main/kotlin/org/eclipse/tractusx/bpdm/orchestrator/service/GoldenRecordTaskService.kt [hunk 1]
```diff
@@ -105,40 +111,25 @@ class GoldenRecordTaskService(
...
-    private fun checkForRetentionTimeouts() {
-        taskRepository.findByProcessingStateRetentionTimeoutBefore(DbTimestamp.now())
-            .forEach {
-                try {
-                    logger.info { "Removing task ${it.uuid} after reaching retention timeout" }
-                    taskRepository.delete(it)
-                } catch (err: RuntimeException) {
-                    logger.error(err) { "Error handling retention timeout for task ${it.uuid}" }
-                }
+    fun processTimeoutTaskInBatches(
```
Comment (nicoprow) [inline bpdm-orchestrator/src/main/kotlin/org/eclipse/tractusx/bpdm/orchestrator/service/GoldenRecordTaskService.kt:115]: better split up into two functions. One for pending, one for retentions

File: bpdm-orchestrator/src/main/kotlin/org/eclipse/tractusx/bpdm/orchestrator/service/TimeoutProcessBatchService.kt [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+            )
+
+            entityManager.clear() // Clear the persistence context to free
+            if (!paginationInfo.hasProcessedTasks) {
+                pageNumber++
+            }
+        } while (paginationInfo.hasNextPage)
+    }
+
+    fun checkForRetentionTimeouts() {
```
Comment (nicoprow) [inline bpdm-orchestrator/src/main/kotlin/org/eclipse/tractusx/bpdm/orchestrator/service/TimeoutProcessBatchService.kt:68]: The two check functions could be merged in one function that takes a golden record task service timeout function as a parameter

**Review by nicoprow — COMMENTED** [review]
Details for my previous comments in the code

**Comment by SujitMBRDI** [comment]
> The logic looks sound to me but I think using such an unspecifc public function in the golden record task service in which you can basically place in any fetch and process function to be executed seems to be quite over the top. Especially when we consider that the TimeoutProcessBatchService does have two very similar methods that just differentiate in between which they give delegates as parameters to that function in the golden record task service.
> 
> Why not just create two separate functions in the golden record task service, one for fetching and processing pending timeouts, one for fetching and processing retention timeouts. If you then want to reduce duplicate code you could create a private function for batch processing these functions.

Yeah, will refactor code accordingly.

**Comment by SujitMBRDI** [comment]
> Details for my previous comments in the code

@nicoprow, i have adapted changes as suggested.

File: bpdm-orchestrator/src/main/kotlin/org/eclipse/tractusx/bpdm/orchestrator/service/TimeoutProcessBatchService.kt [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+            logger.error(err) { "Error checking for timeouts" }
+        }
+    }
+
+    private fun processTimeouts(processFunction: (Int) -> PaginationInfo) {
+        val pageSize = 1000  // Adjust the page size based on memory constraints
+        do {
+            val paginationInfo = processFunction(pageSize)
+            entityManager.clear() // Clear the persistence context to free memory
+        } while (paginationInfo.hasNextPage)
```
Comment (nicoprow) [inline bpdm-orchestrator/src/main/kotlin/org/eclipse/tractusx/bpdm/orchestrator/service/TimeoutProcessBatchService.kt:50]: You could return here the amount of processed tasks

File: bpdm-orchestrator/src/main/kotlin/org/eclipse/tractusx/bpdm/orchestrator/service/TimeoutProcessBatchService.kt [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+
+    @Scheduled(cron = "\${bpdm.task.timeoutCheckCron}")
+    fun processForTimeouts() {
+        try {
+            logger.debug { "Checking for timeouts" }
+            processTimeouts(goldenRecordTaskService::processPendingTimeouts)
+            processTimeouts(goldenRecordTaskService::processRetentionTimeouts)
+        } catch (err: RuntimeException) {
+            logger.error(err) { "Error checking for timeouts" }
+        }
```
Comment (nicoprow) [inline bpdm-orchestrator/src/main/kotlin/org/eclipse/tractusx/bpdm/orchestrator/service/TimeoutProcessBatchService.kt:42]: Here we could have a final log for the total amount of processed tasks and that the scheduling is finished. When testing I was not sure whether the schedule was still going or already finished

**Review by nicoprow — APPROVED** [review]
Looks good to me. Minor suggestions in the comments

File: bpdm-orchestrator/src/main/kotlin/org/eclipse/tractusx/bpdm/orchestrator/service/TimeoutProcessBatchService.kt (hunk 3 above)
Comment (SujitMBRDI) [reply bpdm-orchestrator/src/main/kotlin/org/eclipse/tractusx/bpdm/orchestrator/service/TimeoutProcessBatchService.kt:58]: Adapted changes with PaginationInfo.

File: bpdm-orchestrator/src/main/kotlin/org/eclipse/tractusx/bpdm/orchestrator/service/TimeoutProcessBatchService.kt (hunk 4 above)
Comment (SujitMBRDI) [reply bpdm-orchestrator/src/main/kotlin/org/eclipse/tractusx/bpdm/orchestrator/service/TimeoutProcessBatchService.kt:47]: Done

**Review by nicoprow — APPROVED** [review]

**Pull request merged**